Framework/Algorithms/src/MonteCarloAbsorption.cpp

```diff
@@ -67,6 +81,332 @@ struct EFixedProvider {
+ */
+bool constantEFixed(const EFixedProvider &eFixed,
+                    const std::vector<Mantid::detid_t> &detIDs) {
+  const auto e = eFixed.value(detIDs[0]);
+  for (size_t i = 1; i < detIDs.size(); ++i) {
+    if (e != eFixed.value(detIDs[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/** Creates a rectangular cuboid shape.
+ *  @return A cube shape.
+ */
+Object_sptr makeCubeShape() {
+  using namespace Poco::XML;
+  const double dimension = 0.05;
+  AutoPtr<Document> shapeDescription = new Document;
+  AutoPtr<Element> typeElement = shapeDescription->createElement("type");
+  typeElement->setAttribute("name", "detector");
+  AutoPtr<Element> shapeElement = shapeDescription->createElement("cuboid");
+  shapeElement->setAttribute("id", "cube");
+  const std::string posCoord = std::to_string(dimension / 2);
+  const std::string negCoord = std::to_string(-dimension / 2);
+  AutoPtr<Element> element =
+      shapeDescription->createElement("left-front-bottom-point");
+  element->setAttribute("x", negCoord);
+  element->setAttribute("y", negCoord);
+  element->setAttribute("z", posCoord);
+  shapeElement->appendChild(element);
+  element = shapeDescription->createElement("left-front-top-point");
+  element->setAttribute("x", negCoord);
+  element->setAttribute("y", posCoord);
+  element->setAttribute("z", posCoord);
+  shapeElement->appendChild(element);
+  element = shapeDescription->createElement("left-back-bottom-point");
+  element->setAttribute("x", negCoord);
+  element->setAttribute("y", negCoord);
+  element->setAttribute("z", negCoord);
+  shapeElement->appendChild(element);
+  element = shapeDescription->createElement("right-front-bottom-point");
+  element->setAttribute("x", posCoord);
+  element->setAttribute("y", negCoord);
+  element->setAttribute("z", posCoord);
+  shapeElement->appendChild(element);
+  typeElement->appendChild(shapeElement);
+  AutoPtr<Element> algebraElement = shapeDescription->createElement("algebra");
+  algebraElement->setAttribute("val", "cube");
+  typeElement->appendChild(algebraElement);
+  ShapeFactory shapeFactory;
+  return shapeFactory.createShape(typeElement);
+}
+
+/** Create a workspace whose instrument approximates that of modelWS.
+ *  @param modelWS A workspace to model.
+ *  @param grid An object defining the detector grid of the output workspace.
+ *  @param wavelengthPoints Number of points in the output workspace.
+ *  @return A workspace with sparse instrument.
+ */
+MatrixWorkspace_uptr
+createSparseWS(const MatrixWorkspace &modelWS,
+               const Mantid::Algorithms::DetectorGridDefinition &grid,
+               const size_t wavelengthPoints) {
+  // Build a quite standard and somewhat complete instrument.
+  auto instrument = boost::make_shared<Instrument>("MC_simulation_instrument");
+  instrument->setReferenceFrame(
+      boost::make_shared<ReferenceFrame>(Y, Z, Right, ""));
+  // Add sample to origin.
+  const V3D samplePos{0.0, 0.0, 0.0};
```

Comment:
The original reference frame can be (and should be) taken into account, but I'd rather move the sparse instrument such that the sample position is in the origin as the beam profile stuff kind of expects it.